Solutions for advent of code, day 4 - 13.

zygoloid opened the pull request:
A collection of additional Carbon examples demonstrating current language capabilities and some limitations of the current state of the toolchain.

zygoloid committed:
Solution for advent of code day 4.

zygoloid committed:
Solution for advent of code day 5.

zygoloid committed:
Solution for advent of code day 6.

zygoloid committed:
Solution for advent of code day 7.

zygoloid committed:
Solution for advent of code day 8.

zygoloid committed:
Solution for advent of code day 9.

zygoloid committed:
Solution for advent of code day 10.

zygoloid committed:
Solution for advent of code day 11.

zygoloid committed:
Solution for advent of code day 12.

zygoloid committed:
Solution for advent of code day 13.

jonmeow commented on examples/advent2024/BUILD:
```diff
@@ -9,50 +9,14 @@ utils = [
-        "day3_part2.carbon",
-    ] + utils,
-)
+[
```
```suggestion
# Produce a binary "dayX_partY" for each matching `.carbon` file.
[
```

With the

jonmeow commented on examples/advent2024/BUILD:
```diff
@@ -9,50 +9,14 @@ utils = [
-)
+[
+    carbon_binary(
+        name = "day%d_part%d" % (day, part),
+        srcs = [
+            "day%d_common.carbon" % day,
+            "day%d_part%d.carbon" % (day, part),
```
Had you considered using \.format instead? I note this because I think we're predominantly using it in `.bzl` files (somehow I think this may be the first case in a BUILD)

```suggestion
        name = "day{0}_part{1}".format(day, part),
        srcs = [
            "day{0}_common.carbon".format(day),
            "day{0}_part{1}.carbon".format(day, part),
```

jonmeow commented on examples/advent2024/day10_common.carbon:
```diff
@@ -0,0 +1,27 @@
+import Core library "io";
+import library "io_utils";
+
+class Terrain {
```
Is it worth having any documentation in these files? A lot of these get fairly complex, including when you get into hex representations of characters. Where do you see the balance of maintenance and discarding if the code needs updated? (maybe it's more on the latter side, due to a lack of tests?)

zygoloid replied:
Regarding documentation: [while we are asked not to distribute the test inputs or the problem description](https://adventofcode.com/2024/about#:~:text=Can%20I%20copy/redistribute%20part%20of%20Advent%20of%20Code%3F), I think it'd be fine to include links to the problems. I've gone ahead and done that, but that only helps for part 1 -- the part 2 description is unavailable unless you are logged in and have solved part 1.

I've been a bit wary of including too much description in these because of the request to not distribute the problems, but perhaps I'm being too cautious. I'm intending to do some more maintenance of these examples over time (using new Carbon features as they become available, for example), and will think about striking a better balance here as I rework these.

Regarding maintenance versus discarding: I've added some instructions to `README.md` for how to react to build breaks. I'm leaning towards disabling rather than deleting the examples when they break, but I don't mind much either way; it's easy to get them back from git history if we'd prefer to delete them immediately.

jonmeow replied:
LGTM for the README.md changes (note the possible mistake, separately commented).

The day links  don't help me though -- with the README, I know where to find those. To try to explain, here's how I think I'd give a good review of the code:

- Make up for missing documentation:
  1. Read the advent day.
  2. Think about how I'd solve it, maybe solve myself.
  3. Compare my solution with your code to think about naming and structure.
- Make up for missing tests:
  1. Check out your PR.
  2. Take the advent inputs, and run your PR over the inputs.

Specific reasons here, to try to help you understand:

- `Terrain` isn't mentioned in the day 10 description, though after spending some time trying to review, I think I understand where the _name_ came from.
- `43` I'm lost on, so I'd need to invest more time.
- `0x30` is probably a character, but I don't memorize these things and would need to look it up.

But, that's just one fragment of one day, and I feel I could spend a good few days trying to really understand the code. I understand your desire to avoid comments and tests, but that's also leading to me giving a very light review because I expect it's not worth my time to give deep review.

I don't know if implementing something like https://rosettacode.org/ tasks would be as interesting to you, but that may be an option for things where you could have documentation and tests, possibly making it more approachable to both review and to developers trying to learn about Carbon.

jonmeow reviewed (approved):
Approving, but maybe it's worth a note about what the expectation is (in the README) for the code maintenance is, given the lack of documentation and tests? It says it's mainly for seeing the current state of the toolchain, but what happens if an example stops working? Who fixes that and how do they verify their fix, or do we not fix it and delete instead?

jonmeow commented:
To be sure it's clear, the lack of comments and tests means I can't really offer much review beyond the BUILD beyond "looks like Carbon code".

danakj commented:
> Approving, but maybe it's worth a note about what the expectation is (in the README) for the code maintenance is, given the lack of documentation and tests? It says it's mainly for seeing the current state of the toolchain, but what happens if an example stops working? Who fixes that and how do they verify their fix, or do we not fix it and delete instead?

I was also wondering about this, as we make changes going forward. Maybe it'd be good to clearly timestamp these somehow like "they work with version <githash> of the toolchain" so that we can explicitly not have to maintain them.

zygoloid committed:
Apply suggestions from code review

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

zygoloid committed:
Add a link to the problem being solved. Add instructions to README
explaining what to do if the build breaks, and add a mechanism in the
BUILD file to make it easy.

jonmeow commented on examples/advent2024/day10_common.carbon:
```diff
@@ -0,0 +1,31 @@
+
+// https://adventofcode.com/2024/day/10
+
+// https://adventofcode.com/2024/day/1
```
It looks like you might've accidentally linked to day 1 on days >= 10

zygoloid committed:
Fix action of overenthusiastic script.

mingodad commented:
When trying to parse the examples they are failing because `package` declaration is missing, if the `package` declaration can be omitted then the grammar should be changed to allow it.
Also declaration like `library "day10_common";` is not in the grammar.

jonmeow commented:
IIUC from #4909, you're trying to use the explorer. This is for the toolchain. The explorer codebase is frozen and won't receive new features.
